storage: overhaul GCQueue.shouldQueue

tbg opened the pull request:
For early review. Haven't updated tests yet and want to test drive this on adriatic.

TL;DR: use a scale-invariant `shouldQueue` for the GC Queue to avoid it running
in a tight loop without actually deleting anything. Furthermore don't queue for
GC if the replica has been processed within the TTL to mitigate a similar blow-
up due to the intent GC score.

See #15756

--

When a key of size `B` is deleted at timestamp `T` or superseded by a newer
version, it henceforth is accounted for in the range's `GCBytesAge`. At time
`S`, its contribution to age will be `B*seconds(S-T)`. The aggregate
`GCBytesAge` of all deleted versions in the cluster is what the GC queue at
the time of writing bases its `shouldQueue` method on.

If a replica is queued to have its old values garbage collected, its contents
are scanned. However, the values which are deleted follow a criterion that
isn't immediately connected to `GCBytesAge`: We (basically) delete everything
that's older than the Replica's `TTLSeconds`.

Thus, it's not obvious that garbage collection has the effect of reducing the
metric that we use to consider the replica for the next GC cycle, and it seems
that we messed it up.

The current metric used for queueing: `GCBytesAge/(1<<20 * ttl)` does not
have the right scaling. For example, consider that a value of size `1mb` is
overwritten with a newer version. After `ttl` seconds, it contributes `1mb` to
`GCBytesAge`, and so the replica has a score of `1`, i.e. (roughly) the range
becomes interesting to the GC queue. When GC runs, it will delete value that
are `ttl` old, which our value is. But a Replica is ~64mb, so picture that
you have 64mb of key-value data all at the same timestamp, and they become
superseded. Already after `ttl/64`, the metric becomes 1, but they keys won't
be GC'able for another (63*ttl)/64. Thus, GC will run "all the time" long
before it can actually have an effect.

The metric with correct scaling must thus take into account the size of the
range. What size exactly? Any data that isn't live (i.e. isn't readable by
a scan from the far future). That's `KeyBytes + ms.ValBytes - ms.LiveBytes`,
which is also known as `GCBytes` in the code. Hence, the better metric is
`GCBytesAge/(ttl*GCBytes)`.

Using this metric guarantees that after truncation, `GCBytesAge` is at most
`ttl*GCBytes` (where `GCBytes` has been updated), i.e. the new metric is at
most 1.

To visualize this, picture a rectangular frame of width `ttl` and height
`GCBytes` (i.e. the horizontal dimension is time, the vertical one bytes),
where the right boundary of the frame corresponds to age zero.
Each non-live key is a domino aligned with the right side of the frame, its
width equal to its size, and its height given by the duration (in seconds) it's
been non-live.

The combined surface of the dominos is then `GCBytesAge`, and the claim is that
if the total sum of domino heights (i.e. sizes) is `GCBytes`, and the surface
is larger than `ttl*GCBytes` by some positive `X`, then after removing the
dominos that cross the line `x=-ttl` (i.e. `ttl` to the left from the right
side of the frame), at least a surface area of `X` has been removed.

```           x=-ttl                 GCBytes=1+4
                |           3 (age)
                |          +-------+
                |          | keep  | 1 (bytes)
                |          +-------+
           +-----------------------+
           |                       |
           |        remove         | 3 (bytes)
           |                       |
           +-----------------------+
                |   7 (age)
```

This is true because

```
deletable area  = total area       - nondeletable area
                = X + ttl*GCBytes  - nondeletable area
               >= X + ttl*GCBytes  - ttl*(bytes in nondeletable area)
                = X + ttl*(GCBytes - bytes in nondeletable area)
               >= X.
```

Or, in other words, you can only hope to put `ttl*GCBytes` of area in the
"safe" rectangle. Once you've done that, everything else you put is going to be
deleted.

This means that running GC will always result in a `GCBytesAge` of `<=
ttl*GCBytes`, and that a decent trigger for GC is a multiple of `ttl*GCBytes`.

Limitations:

- Queueing for GC is nevertheless expensive due to the full scan involved, even
  if it gets to delete data. Small TTLs pose an obvious problem, and those
  should be throttled (i.e. bump a small TTL to a reasonable one implicitly).
- Replicas with heavy continuous deletion workloads could also pose a problem
  since the amount of data amassed on the Replica before GC kicks in could be
  prohibitive. There are likely scenarios in which it is more reasonable to
  perform garbage collection as the keys are accessed, though not even that can
  cover the scenario in which keys are never overwritten repeatedly. If it
  should come to that, the only reasonable path forward is excluding
  sufficiently outdated data from the consistent part of the replica, which
  excludes it from further replication (snapshots) and allows ad-hoc cleanup.
- `ttl=0` is particularly interesting in this regard since historical versions
  can always be deleted instantly and so GC runs can be driven directly by the
  number of non-live bytes.
- The intent portion of `shouldQueue` could be another time bomb if it fires
  continuously for some reason.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15897)
<!-- Reviewable:end -->

BramGruneir commented:
This updated limit sounds a lot better and I love the ascii visuals.

Would it be possible to cut the scan short if it passes some size threshold for deletion?  Say 1mb?  And perform the deletions in a more piecemeal way?

What about adding adding an intent resolution queue instead of handling it in the gcqueue?  I've been scanning through the code for the gcqueue and it's already very complex, by pulling that out into another queue, it would make it much cleaner.  And when the gcqueue detected an intent, it could send enqueue it into the other queue.

---

Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/storage/gc_queue.go, line 135 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-Kjxr13GYcH6YDwGZWlM:-Kjxr13GYcH6YDwGZWlN:bdlcjer) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4c98f7c60698f7b8eb26875e160c9959a595d59/pkg/storage/gc_queue.go#L135)):*
> ```Go
> 	repl.mu.Unlock()
> 
> 	ms := repl.GetMVCCStats()
> ```

Can you just keep the lock and grab r.mu.state.Stats directly?  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-:-KjxtcAcRErAKGPKEaJd:b-5tisju)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
> Would it be possible to cut the scan short if it passes some size threshold for deletion? Say 1mb? And perform the deletions in a more piecemeal way?

Maybe the PR message was a bit too rambly: I'm worried that now that the criterion for queueing scales with the replica size, that we can end up having replicas that are 99% garbage and yet, all that garbage would be sent along with snapshots, etc.

We should also, and I think that is what you suggest, send smaller batches for the actual `GCRequest`s. That's fairly straightforward and I'll get to that in a later PR. If you wanted to limit how much we actually scan, you'd have to add state to the replica so that a later queue run would know where to "restart". I'd rather not do that until there's good benefit to it. We're reading from a snapshot, so we shouldn't get in the way of things too much, though we may consider throttling max kv/sec at some point.

> What about adding adding an intent resolution queue instead of handling it in the gcqueue?

Definitely agree that the GCQueue tries to do a lot and is fairly complex, and once intent GC actually happens on some practical workloads and we have to touch it, it's possible that that's what we're going to do. For now, I'll leave the intent and transaction handling alone.

tschottdorf commented:
Initial testing on `adriatic` looks decent. While running (and having run) `kv` loadgens, I set the GC TTL from default (1 day) to 10 minutes. This triggered extensive GC (which bogged down the system for a while) but once it was through, the later GC activity seems much more measured. Traces show that most of the time (seconds) is spent in Raft. Note that the load generators put the nodes under a lot of stress even without GC - consequently the GCRequests impact quite negatively on an already overloaded system, with a temporary drop in throughput.

![image](https://cloud.githubusercontent.com/assets/5076964/26020661/5a64370e-374f-11e7-87c1-ebf0ed10de5b.png)

tschottdorf commented on pkg/storage/gc_queue.go:
```diff
@@ -125,25 +123,40 @@ func (gcq *gcQueue) shouldQueue(
-	gcScore := float64(ms.GCByteAge(now.WallTime)) / float64(zone.GC.TTLSeconds) / float64(gcByteCountNormalization)
+	// GC score is the total GC'able bytes age normalized by (replica size * the replica's TTL in seconds).
+	// TODO(tschottdorf): ttlSeconds = 0 || ValBytes == 0?
+	denominator := float64(ttlSeconds * ms.ValBytes)
```
actually this should be `GCBytes()`. You're probably best off just looking at the PR description for now. I'm also going to have to do something about the case in which there isn't much to GC, to make sure we don't even bother to GC in that case.

tschottdorf commented:
Current version. I dropped the requirement to process a replica only once per TTL as it shouldn't be necessary (at least when queueing based on key size). This is on `adriatic` with each node running `kv` with 10k cycle length and <1kb writes, with around 300 ranges in the system. GC cycles during load remove on the order of 8mb of data. After load stops, it takes a few more hours until everything is GC'ed, but each cycle does do "real" work as expected. There's some more work to do the edge cases (lot of data, not so much garbage - expensive to run GC on) but I'll do that through unit testing. No need to review this just yet.

![image](https://cloud.githubusercontent.com/assets/5076964/26040029/709df306-38ef-11e7-92d5-31ea5c570b75.png)

tamird commented:
Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KjyAsRDDcJ0JpL5POS3:-KjyAsREjnAOKpBHV9jR:btmauxh) ([raw file](https://github.com/cockroachdb/cockroach/blob/80d7745ac3179b34b0937a9b278e52570fb900ac/pkg/storage/gc_queue.go#L132)):*
> ```Go
> 
> 	repl.mu.Lock()
> 	considerGC := now.Less(repl.mu.state.GCThreshold.Add((ttlSeconds*1E9)/2, 0))
> ```

nit: avoid slightly magic numbers by writing this as `.Add((ttlSeconds*time.Second).Nanoseconds()/2, 0)`

---

*[pkg/storage/replica_state.go, line 106 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KjyA59tRdGSLLQGsATY:-KjyA59tRdGSLLQGsATZ:bh1rgvr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3c50e3b2debde49a085405f8e5031cc8c2b0796/pkg/storage/replica_state.go#L106)):*
> ```Go
> //
> // TODO(tschottdorf): test and assert that none of the optional values are
> // missing when- ever saveState is called. Optional values should be reserved
> ```

might take this occasion to rename-proof this by avoiding the method name.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897)*
<!-- Sent from Reviewable.io -->

petermattis commented:
That is an epic commit message!

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-:-KkCbPae74GvqOZKDZV6:b-6yhiw7)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Still working on the unit tests, but I'm now happy with the way this works for my `adriatic` cycle-len 10k workload (note the dispersion on GC activity; I added slight fuzzing).

![image](https://cloud.githubusercontent.com/assets/5076964/26124517/bcb855b6-3a4c-11e7-9b9c-e8cea8c7db33.png)

Hopefully reviewable tomorrow or day after.

tschottdorf commented:
Review status: 3 of 8 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/storage/gc_queue.go, line 132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KjyAsRDDcJ0JpL5POS3:-KkM0uZieunmvp8VRc2-:b-b3jbv4) ([raw file](https://github.com/cockroachdb/cockroach/blob/80d7745ac3179b34b0937a9b278e52570fb900ac/pkg/storage/gc_queue.go#L132)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: avoid slightly magic numbers by writing this as `.Add((ttlSeconds*time.Second).Nanoseconds()/2, 0)`
</blockquote></details>

Obsolete.

---

*[pkg/storage/gc_queue.go, line 135 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-Kjxr13GYcH6YDwGZWlM:-KkM0w3C_hVyr6Bnyenm:b-ngcfp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/80d7745ac3179b34b0937a9b278e52570fb900ac/pkg/storage/gc_queue.go#L135)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Can you just keep the lock and grab r.mu.state.Stats directly?  
</blockquote></details>

Should be done, though the code moved.

---

*[pkg/storage/replica_state.go, line 106 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KjyA59tRdGSLLQGsATY:-KkM0ya5W5aqWivh-6QC:b9vd9bw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3c50e3b2debde49a085405f8e5031cc8c2b0796/pkg/storage/replica_state.go#L106)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

might take this occasion to rename-proof this by avoiding the method name.
</blockquote></details>

I already feel bad for the random edit, will leave it at that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Ready for review.

tamird commented:
Reviewed 3 of 5 files at r2.
Review status: 6 of 8 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_state.go, line 106 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KjyA59tRdGSLLQGsATY:-KkM2VTjPzzKAThVpMco:b-dbrg7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3c50e3b2debde49a085405f8e5031cc8c2b0796/pkg/storage/replica_state.go#L106)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I already feel bad for the random edit, will leave it at that.
</blockquote></details>

You're setting yourself up for more future random edits.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 7 files at r1, 3 of 5 files at r2, 2 of 2 files at r3.
Review status: 7 of 8 files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[pkg/storage/gc_queue.go, line 117 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkM8yMpM3fMDSuiyDHH:-KkM8yMpM3fMDSuiyDHI:bfqak5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L117)):*
> ```Go
> type resolveFunc func([]roachpb.Intent, bool, bool) error
> 
> type gcQueueScore struct {
> ```

Add a comment that these fields are documented where they are assigned in shouldQueueImpl, and that this struct exists mainly to make the internal operations of shouldQueueImpl testable. 

---

*[pkg/storage/gc_queue.go, line 119 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMB5eF9fJaxHGJsLc8:-KkMB5eF9fJaxHGJsLc9:b-g81ubn) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L119)):*
> ```Go
> type gcQueueScore struct {
> 	TTL                 time.Duration
> 	LikelyLastGC        time.Duration
> ```

LikelyLastGC is not commented anywhere. Since it's only used in tests, maybe it should just be computed in the tests instead of in the real code?

---

*[pkg/storage/gc_queue.go, line 132 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkM9K2BJJvD_F1M_kF8:-KkM9K2BJJvD_F1M_kF9:bn39ctk) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L132)):*
> ```Go
> }
> 
> func (r gcQueueScore) ValuesScore() float64 {
> ```

It's weird that this is factored out into a method even though it's only used once and everything else is just inlined. 

---

*[pkg/storage/gc_queue.go, line 165 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMCoQf8riqo3xAMNw-:-KkMCoQgclVZIXlFzc_t:b-vw07zi) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L165)):*
> ```Go
> ) gcQueueScore {
> 	repl.mu.Lock()
> 	ms := *(&repl.mu.state.Stats)
> ```

This can just be `ms := repl.mu.state.Stats`

---

*[pkg/storage/gc_queue.go, line 185 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkM9mNfWW_XIczOYP5g:-KkM9mNg1lB0uBK6rz43:buknw93) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L185)):*
> ```Go
> }
> 
> func shouldQueueImpl(
> ```

Rename this to replicaToScoreImpl; it's used in more than just shouldQueue now.

Consider including most of the commit message as a comment in the code to make it more visible in the future.

---

*[pkg/storage/gc_queue.go, line 195 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMAjHNExO3n4bo1B6A:-KkMAjHNExO3n4bo1B6B:b-nslv6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L195)):*
> ```Go
> 		// There is no real-world difference between 0 and 1, but 1 is better for math.
> 		// TODO(tschottdorf): this edge case may not even be possible. Should not FUD.
> 		r.TTL = time.Second
> ```

If you're just trying to avoid zero, why `time.Second` instead of `time.Duration(1)`? Or you could modify `ttlSeconds` before `r.TTL` is set so it's clear that everything stays in integer seconds. (is r.TTL even necessary? you're just doing `.Seconds()` on it every time you use it here)

---

*[pkg/storage/gc_queue.go, line 209 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMDIYprga7sBI9Bc1p:-KkMDIYprga7sBI9Bc1q:b-f9pry8) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L209)):*
> ```Go
> 	// values, this should be (almost) 1. For one just hit completely by a DeleteRange,
> 	// it should be (almost) zero.
> 	r.LiveFraction = float64(ms.LiveBytes) / float64(1+ms.ValBytes+ms.KeyBytes) // +1 avoids NaN
> ```

Can the stats go negative when ms.ContainEstimates is true? We may need explicit checks for `<= 0` instead of just adding one. 

---

*[pkg/storage/gc_queue_test.go, line 123 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMEeZCM-dJrMWfUK69:-KkMEeZCM-dJrMWfUK6A:b-eap8iv) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue_test.go#L123)):*
> ```Go
> const cacheFirstLen = 3
> 
> type cachee struct {
> ```

s/cachee/gcTestCacheKey/

---

*[pkg/storage/gc_queue_test.go, line 127 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMEnq2MdJH2x2nipmT:-KkMEnq2MdJH2x2nipmU:b-lw2971) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue_test.go#L127)):*
> ```Go
> 	string
> }
> type cacheVal struct {
> ```

s/cacheVal/gcTestCacheVal/

---

*[pkg/storage/gc_queue_test.go, line 142 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMFEdY0g9hTQXM82Hp:-KkMFEdY0g9hTQXM82Hq:b-sd60ez) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue_test.go#L142)):*
> ```Go
> 	cws.cache = map[cachee]cacheVal{
> 		{enginepb.MVCCStats{LastUpdateNanos: 946684800000000000}, "1-1m0s-1.0 MiB"}: {
> 			first: [3]enginepb.MVCCStats{
> ```

s/3/cacheFirstLen/

---

*[pkg/storage/gc_queue_test.go, line 260 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMFYxvOTQj5uGuNi_P:-KkMFYxvOTQj5uGuNi_Q:b-pkiimm) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue_test.go#L260)):*
> ```Go
> 
> 	{
> 		// Hammer a key with 1MB blobs for a minute. Logically, at the end of that
> ```

s/minute/(simulated) minute/

---

*[pkg/storage/engine/enginepb/mvcc.go, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-Kk7zgNZY8F77HgygtZt:-Kk7zgNZY8F77HgygtZu:b8b25vt) ([raw file](https://github.com/cockroachdb/cockroach/blob/80d7745ac3179b34b0937a9b278e52570fb900ac/pkg/storage/engine/enginepb/mvcc.go#L70)):*
> ```Go
> 		return
> 	}
> 	diffSeconds := nowNanos/1E9 - ms.LastUpdateNanos/1E9 // not (a-b)/1E9!
> ```

Why not?

---

*[pkg/storage/engine/enginepb/mvcc.proto, line 133 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkM8ncYz1Bq99fw4nbq:-KkM8ncZG2Urn2NdgxDS:bh3jhgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/engine/enginepb/mvcc.proto#L133)):*
> ```protobuf
>   optional sfixed64 gc_bytes_age = 3 [(gogoproto.nullable) = false, (gogoproto.customname) = "GCBytesAge"];
>   // live_bytes is the number of bytes stored in keys and values which can in
>   // principle be read by means of a Scan or Get "at infinity", including
> ```

s/at infinity/in the far future/ to make it clear we're talking about time here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-:-KkMFn4w-g0AhpTmq1fB:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 7 of 8 files reviewed at latest revision, 20 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 120 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkM9SYvIMrWpv-uuLbT:-KkM9SYvIMrWpv-uuLbU:b-chatgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L120)):*
> ```Go
> 	TTL                 time.Duration
> 	LikelyLastGC        time.Duration
> 	LiveFraction        float64
> ```

Should this be `DeadFraction`? You use `1 - LiveFraction` in at least 2 places.

---

*[pkg/storage/gc_queue.go, line 143 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkM9CfcCmiznyubg4A7:-KkM9CfdzOqUfaViWwJK:b-bieebv) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L143)):*
> ```Go
> 		r.ExpMinGCByteAgeReduction = 0
> 	}
> 	return fmt.Sprintf("queue=%t with %.2f/fuzz(%.2f)=%.2f=valScalScore(%.2f)*deadFrac(%.2f)+intentScore(%.2f)\n"+
> ```

nit: `s/ScalScore/ScaleScore/g` (I found the abbreviation `Scal` jarring).

---

*[pkg/storage/gc_queue.go, line 161 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkM9fp0nXOoZ8VY3N6t:-KkM9fp0nXOoZ8VY3N6u:b-qmj68o) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L161)):*
> ```Go
> }
> 
> func replicaToScore(
> ```

Can we make the name more specific to the GC queue: `makeGCQueueScore`?

The interaction between `gcQueue.shouldQueue`, `replicaToScore` and `shouldQueueImpl` is interesting. Guessing this was for testability.

---

*[pkg/storage/gc_queue.go, line 224 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMBcnjFMLP5qLE_0Yp:-KkMBcnjFMLP5qLE_0Yq:b-5oaklh) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L224)):*
> ```Go
> 	// Intent score. This computes the average age of outstanding intents and
> 	// normalizes. Note that at the time of writing this criterion hasn't
> 	// undergone a reality check yet.
> ```

Should this be a TODO instead of a Note?

---

*[pkg/storage/gc_queue.go, line 448 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMGZYgWTj5QqIHT-ec:-KkMGZYgWTj5QqIHT-ed:b-9uxw6s) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L448)):*
> ```Go
> 	}
> 
> 	info.BeforeMVCCStats = repl.GetMVCCStats()
> ```

Are you using `BeforeMVCCStats`? I'm not seeing it.

---

*[pkg/storage/engine/enginepb/mvcc.go, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-Kk7zgNZY8F77HgygtZt:-KkMH33qXXvBq7UAWj8q:bgm2o8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/80d7745ac3179b34b0937a9b278e52570fb900ac/pkg/storage/engine/enginepb/mvcc.go#L70)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why not?
</blockquote></details>

I recall there were rounding differences. Should definitely immortalize the reason in the comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-:-KkMHGEMwh606648tow5:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Updated, PTAL.

---

Review status: 4 of 8 files reviewed at latest revision, 20 unresolved discussions.

---

*[pkg/storage/gc_queue.go, line 117 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkM8yMpM3fMDSuiyDHH:-KkMT71_YlGUqS_eG1l9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L117)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Add a comment that these fields are documented where they are assigned in shouldQueueImpl, and that this struct exists mainly to make the internal operations of shouldQueueImpl testable. 
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 119 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMB5eF9fJaxHGJsLc8:-KkMT8c-UZ7xNNxNkfJ1:bhk26lv) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L119)):*
It's output with the string representation, which goes to tracing (and that's the main reason it's here - I want to look at a GC trace and know everything).

> [gc,n5,s5,r37/3:/Table/51/1/2{867862…-930850…}] processing replica with score 2.12/fuzz(1.06) = valScalScore(2.34)*deadFrac(0.86)+intentScore(0.00) = 2.01 likely last GC: 34m27.000168071s ago, 7.2 MiB non-live, curr. age 9.8 GiB*s, min exp. reduction: 5.6 GiB*s

---

*[pkg/storage/gc_queue.go, line 120 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkM9SYvIMrWpv-uuLbT:-KkMUgsh3QN5lZDwO5kd:b-k816ve) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L120)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Should this be `DeadFraction`? You use `1 - LiveFraction` in at least 2 places.
</blockquote></details>

Done. RIP LiveFraction, long live DeadFraction.

---

*[pkg/storage/gc_queue.go, line 132 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkM9K2BJJvD_F1M_kF8:-KkMTXHuozVsFD5f3SWh:b-q5v9nj) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L132)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It's weird that this is factored out into a method even though it's only used once and everything else is just inlined. 
</blockquote></details>

Do you feel strongly? It happens to be a function of two other fields, and while I didn't fully duplicate the scores, I'd do that first rather than add duplication.

---

*[pkg/storage/gc_queue.go, line 143 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkM9CfcCmiznyubg4A7:-KkMTsW8KUfjKIS5_a5s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L143)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

nit: `s/ScalScore/ScaleScore/g` (I found the abbreviation `Scal` jarring).
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 161 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkM9fp0nXOoZ8VY3N6t:-KkMUleUukKeN6mePdI7:bo6wrae) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L161)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Can we make the name more specific to the GC queue: `makeGCQueueScore`?

The interaction between `gcQueue.shouldQueue`, `replicaToScore` and `shouldQueueImpl` is interesting. Guessing this was for testability.
</blockquote></details>

Yep, testability and being able to get a score in traces where it's useful. Done.

---

*[pkg/storage/gc_queue.go, line 165 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMCoQf8riqo3xAMNw-:-KkMUuacztCEpahM-hPN:b-p9a4r7) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L165)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This can just be `ms := repl.mu.state.Stats`
</blockquote></details>

Not sure where that came from. Done.

---

*[pkg/storage/gc_queue.go, line 185 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkM9mNfWW_XIczOYP5g:-KkMVIrG4FDsfiGbL00w:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L185)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Rename this to replicaToScoreImpl; it's used in more than just shouldQueue now.

Consider including most of the commit message as a comment in the code to make it more visible in the future.
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 195 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMAjHNExO3n4bo1B6A:-KkMOWiyYahr7si-_MBC:bw11ywh) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L195)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

If you're just trying to avoid zero, why `time.Second` instead of `time.Duration(1)`? Or you could modify `ttlSeconds` before `r.TTL` is set so it's clear that everything stays in integer seconds. (is r.TTL even necessary? you're just doing `.Seconds()` on it every time you use it here)
</blockquote></details>

TTL is configured in seconds, so it makes more sense to me to treat zero as a one. Besides, the GC queue scanner interval is on the order of seconds, so any smaller value wouldn't have a real effect.
`r.TTL` is returned to the caller to inform them about the TTL on which the priority is based, and by looking at a `time.Duration` I know what it is, whereas an `int` that stores seconds is less intuitive.

I just checked that it is in fact possible to set a zero TTL. I'll wait for your reply before I write a unit test.

---

*[pkg/storage/gc_queue.go, line 209 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMDIYprga7sBI9Bc1p:-KkMXLL20w-_NiweYnAQ:bwhjn7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L209)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Can the stats go negative when ms.ContainEstimates is true? We may need explicit checks for `<= 0` instead of just adding one. 
</blockquote></details>

Good point, I made this more robust - PTAL before I write unit tests.

---

*[pkg/storage/gc_queue.go, line 224 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMBcnjFMLP5qLE_0Yp:-KkM_OhASgeRwLrmvPVS:b-7p5csk) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L224)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Should this be a TODO instead of a Note?
</blockquote></details>

It could be, but I'll just file an issue: https://github.com/cockroachdb/cockroach/issues/15997
Just wanted to make clear that while the code looks polished, the reader would waste their time here if they searched for a deep reason.

---

*[pkg/storage/gc_queue.go, line 448 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMGZYgWTj5QqIHT-ec:-KkMPQBvUNeJMYI_JqmC:b-9kddu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L448)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Are you using `BeforeMVCCStats`? I'm not seeing it.
</blockquote></details>

`info` is printed below, which contains the stats. Removed the field and printing them directly instead.

---

*[pkg/storage/gc_queue_test.go, line 123 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMEeZCM-dJrMWfUK69:-KkMQAO6jQxbgG4NaETO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue_test.go#L123)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/cachee/gcTestCacheKey/
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue_test.go, line 127 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMEnq2MdJH2x2nipmT:-KkMQ6ZvFzCsogqXXAD0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue_test.go#L127)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/cacheVal/gcTestCacheVal/
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue_test.go, line 142 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMFEdY0g9hTQXM82Hp:-KkMQEx9Ot87ZDqw13qR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue_test.go#L142)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/3/cacheFirstLen/
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue_test.go, line 260 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMFYxvOTQj5uGuNi_P:-KkMQJC6yIM_Bed1kVDU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue_test.go#L260)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/minute/(simulated) minute/
</blockquote></details>

Done.

---

*[pkg/storage/engine/enginepb/mvcc.go, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-Kk7zgNZY8F77HgygtZt:-KkMQXpSRotPJtIEg17a:b6p1i2u) ([raw file](https://github.com/cockroachdb/cockroach/blob/80d7745ac3179b34b0937a9b278e52570fb900ac/pkg/storage/engine/enginepb/mvcc.go#L70)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I recall there were rounding differences. Should definitely immortalize the reason in the comment.
</blockquote></details>

Immortalized.

---

*[pkg/storage/engine/enginepb/mvcc.proto, line 133 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkM8ncYz1Bq99fw4nbq:-KkMR76xYzkMfZaFA5lQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/engine/enginepb/mvcc.proto#L133)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/at infinity/in the far future/ to make it clear we're talking about time here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-:-KkMaF9ejztdryAkIvuZ:b-rr168b)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 3 of 8 files reviewed at latest revision, 16 unresolved discussions.

---

*[pkg/storage/gc_queue.go, line 275 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMfjED3a4o1OZkz2HH:-KkMfjED3a4o1OZkz2HI:b-aqijb1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8904a5b05ab42959a3b69ec2604c393e312eda1b/pkg/storage/gc_queue.go#L275)):*
> ```Go
> // This means that running GC will always result in a `GCBytesAge` of `<=
> // ttl*GCBytes`, and that a decent trigger for GC is a multiple of
> // `ttl*GCBytes`.
> ```

This comment is awesome.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 4 of 5 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/storage/gc_queue.go, line 132 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkM9K2BJJvD_F1M_kF8:-KkMkOxk2SuVomnC-90e:b-d4rg13) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L132)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Do you feel strongly? It happens to be a function of two other fields, and while I didn't fully duplicate the scores, I'd do that first rather than add duplication.
</blockquote></details>

What duplication? I was just thinking of a local variable `valuesScore := r.DeadFraction * r.ValuesScalableScore`?

(also: `r` for the receiver name?)

---

*[pkg/storage/gc_queue.go, line 195 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMAjHNExO3n4bo1B6A:-KkMlYCR3tf_XTUJj_C1:bu3kjic) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L195)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

TTL is configured in seconds, so it makes more sense to me to treat zero as a one. Besides, the GC queue scanner interval is on the order of seconds, so any smaller value wouldn't have a real effect.
`r.TTL` is returned to the caller to inform them about the TTL on which the priority is based, and by looking at a `time.Duration` I know what it is, whereas an `int` that stores seconds is less intuitive.

I just checked that it is in fact possible to set a zero TTL. I'll wait for your reply before I write a unit test.
</blockquote></details>

Returning it as a time.Duration is fine; I was just suggesting that you could keep using `ttlSeconds` for all the computation in this message. Either way is OK with me. 

---

*[pkg/storage/gc_queue.go, line 209 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMDIYprga7sBI9Bc1p:-KkMls6UgmAWZ8wV0tZL:b-k8axox) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L209)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Good point, I made this more robust - PTAL before I write unit tests.
</blockquote></details>

`abs()` doesn't seem right - I think we want to clamp negative values at zero instead of inverting them. 

---

*[pkg/storage/gc_queue.go, line 119 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMkePRsa_652G0PfY_:-KkMkePRsa_652G0PfYa:b3mq5de) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1eb892a7f465fc99ab5c75d1627c9330814bda/pkg/storage/gc_queue.go#L119)):*
> ```Go
> // gcQueueScore holds details about the score returned by makeGCQueueScoreImpl for
> // testing and logging. The fields in this struct are documented in
> // shouldQueueImpl.
> ```

s/shouldQueueImpl/makeGCQueueScoreImpl/

---

*[pkg/storage/gc_queue.go, line 305 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMm5FqnjWRpVVwCcnH:-KkMm5FqnjWRpVVwCcnI:b1n63qj) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1eb892a7f465fc99ab5c75d1627c9330814bda/pkg/storage/gc_queue.go#L305)):*
> ```Go
> 	lo, hi := 0.0, 1.0
> 	if ms.ContainsEstimates {
> 		lo, hi = 0.1, 0.9
> ```

Why is an adjustment necessary when there are estimates? How were the values 0.1 and 0.9 determined?

---

*[pkg/storage/gc_queue_test.go, line 127 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMEnq2MdJH2x2nipmT:-KkMkrIre1wU4IUX4lPV:bb4nqwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue_test.go#L127)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Doesn't look done to me. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
One more PTAL for the road.

---

Review status: 6 of 8 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/storage/gc_queue.go, line 132 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkM9K2BJJvD_F1M_kF8:-KkMod99vG08YsLQfnv5:b-tnsttg) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L132)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

What duplication? I was just thinking of a local variable `valuesScore := r.DeadFraction * r.ValuesScalableScore`?

(also: `r` for the receiver name?)
</blockquote></details>

Ah - I thought I called this in the string output as well, which I didn't. Inlined as you suggested.

---

*[pkg/storage/gc_queue.go, line 209 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMDIYprga7sBI9Bc1p:-KkMsCKQojAEh1OtBijS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue.go#L209)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`abs()` doesn't seem right - I think we want to clamp negative values at zero instead of inverting them. 
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 119 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMkePRsa_652G0PfY_:-KkMsL8-z-yE0mqkSyGJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1eb892a7f465fc99ab5c75d1627c9330814bda/pkg/storage/gc_queue.go#L119)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/shouldQueueImpl/makeGCQueueScoreImpl/
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 305 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMm5FqnjWRpVVwCcnH:-KkMsLrpdv9W58XbzGEG:b-v1iq1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1eb892a7f465fc99ab5c75d1627c9330814bda/pkg/storage/gc_queue.go#L305)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why is an adjustment necessary when there are estimates? How were the values 0.1 and 0.9 determined?
</blockquote></details>

I was scared to end up at 0 and never GC and picked safe values. But removed this again, PTAL. The estimates we're seeing in practice should be fairly benign since they only ever come from time series (right?) and if weird things did happen, this obfuscation wouldn't necessarily help us.

---

*[pkg/storage/gc_queue_test.go, line 127 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMEnq2MdJH2x2nipmT:-KkMoaafdU2tMieZuWiY:b-z1r7gt) ([raw file](https://github.com/cockroachdb/cockroach/blob/687823a3adef85a48ee6554a26b188d333c09cec/pkg/storage/gc_queue_test.go#L127)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Doesn't look done to me. 
</blockquote></details>

Ok, another attempt.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-:-KkMt-RY2L8rDllnLwJQ:brmplkz)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: 6 of 8 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/storage/gc_queue.go, line 322 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMvjIA-vE_D55zaQC4:-KkMvjIBMi2gP9e0xfFp:b-t7ogsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/8749139b301e69546b76b9592adc83bac604c739/pkg/storage/gc_queue.go#L322)):*
> ```Go
> 	// needs to be scanned in order to GC. We don't want to run this costly scan
> 	// unless we get a corresponding expected reduction in GCByteAge, so we'll
> 	// use ValuesScore() (which weighs in the fraction of live data) below.
> ```

ValuesScore() no longer exists.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-:-KkMvppdiBvIOPKTt5m9:bb6ywmx)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: overhaul GCQueue.shouldQueue

TL;DR: use a scale-invariant `shouldQueue` for the GC Queue to avoid it running
in a tight loop without actually deleting anything. Adjust this scalable score
so that GC is delayed when a large proportion of the data in the replica is
live. Slightly perturb returned scores to avoid groups of replicas becoming
eligible for GC at the same time repeatedly.

Closes #15897.

--

When a key of size `B` is deleted at timestamp `T` or superseded by a newer
version, it henceforth is accounted for in the range's `GCBytesAge`. At time
`S`, its contribution to age will be `B*seconds(S-T)`. The aggregate
`GCBytesAge` of all deleted versions in the cluster is what the GC queue at
the time of writing bases its `shouldQueue` method on.

If a replica is queued to have its old values garbage collected, its contents
are scanned. However, the values which are deleted follow a criterion that
isn't immediately connected to `GCBytesAge`: We (basically) delete everything
that's older than the Replica's `TTLSeconds`.

Thus, it's not obvious that garbage collection has the effect of reducing the
metric that we use to consider the replica for the next GC cycle, and it seems
that we messed it up.

The previous metric used for queueing: `GCBytesAge/(1<<20 * ttl)` does not
have the right scaling. For example, consider that a value of size `1mb` is
overwritten with a newer version. After `ttl` seconds, it contributes `1mb` to
`GCBytesAge`, and so the replica has a score of `1`, i.e. (roughly) the range
becomes interesting to the GC queue. When GC runs, it will delete value that
are `ttl` old, which our value is. But a Replica is ~64mb, so picture that
you have 64mb of key-value data all at the same timestamp, and they become
superseded. Already after `ttl/64`, the metric becomes 1, but they keys won't
be GC'able for another (63*ttl)/64. Thus, GC will run "all the time" long
before it can actually have an effect.

The metric with correct scaling must thus take into account the size of the
range. What size exactly? Any data that isn't live (i.e. isn't readable by
a scan from the far future). That's `KeyBytes + ms.ValBytes - ms.LiveBytes`,
which is also known as `GCBytes` in the code. Hence, the better metric is
`GCBytesAge/(ttl*GCBytes)`.

Using this metric guarantees that after truncation, `GCBytesAge` is at most
`ttl*GCBytes` (where `GCBytes` has been updated), i.e. the new metric is at
most 1.

To visualize this, picture a rectangular frame of width `ttl` and height
`GCBytes` (i.e. the horizontal dimension is time, the vertical one bytes),
where the right boundary of the frame corresponds to age zero.
Each non-live key is a domino aligned with the right side of the frame, its
width equal to its size, and its height given by the duration (in seconds) it's
been non-live.

The combined surface of the dominos is then `GCBytesAge`, and the claim is that
if the total sum of domino heights (i.e. sizes) is `GCBytes`, and the surface
is larger than `ttl*GCBytes` by some positive `X`, then after removing the
dominos that cross the line `x=-ttl` (i.e. `ttl` to the left from the right
side of the frame), at least a surface area of `X` has been removed.

```           x=-ttl                 GCBytes=1+4
                |           3 (age)
                |          +-------+
                |          | keep  | 1 (bytes)
                |          +-------+
           +-----------------------+
           |                       |
           |        remove         | 3 (bytes)
           |                       |
           +-----------------------+
                |   7 (age)
```

This is true because

```
deletable area  = total area       - nondeletable area
                = X + ttl*GCBytes  - nondeletable area
               >= X + ttl*GCBytes  - ttl*(bytes in nondeletable area)
                = X + ttl*(GCBytes - bytes in nondeletable area)
               >= X.
```

Or, in other words, you can only hope to put `ttl*GCBytes` of area in the
"safe" rectangle. Once you've done that, everything else you put is going to be
deleted.

This means that running GC will always result in a `GCBytesAge` of `<=
ttl*GCBytes`, and that a decent trigger for GC is a multiple of `ttl*GCBytes`.

Limitations:

- The intent portion of `shouldQueue` could be another time bomb if it fires
  continuously for some reason.

tschottdorf commented:
Review status: 5 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/storage/gc_queue.go, line 322 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897#-KkMvjIA-vE_D55zaQC4:-KkMwKxLY9aDwV9ixO2F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8749139b301e69546b76b9592adc83bac604c739/pkg/storage/gc_queue.go#L322)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

ValuesScore() no longer exists.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15897)*
<!-- Sent from Reviewable.io -->
